src/sentry/sentry_metrics/indexer/cloudspanner/cloudspanner.py

```diff
@@ -60,28 +109,283 @@ def validate(self) -> None:
+        using insert_individual_records.
+        """
+
+        def insert_batch_transaction_uow(transaction: Any) -> None:
+            transaction.insert(
+                table=self._table_name, columns=get_column_names(), values=rows_to_insert
+            )
+
+        try:
+            self.database.run_in_transaction(insert_batch_transaction_uow)
+        except google.api_core.exceptions.AlreadyExists as exc:
+            if exc.code == HTTPStatus.CONFLICT:
+                raise CloudSpannerRowAlreadyExists
+        else:
+            metrics.incr(
+                _INDEXER_DB_INSERT_METRIC,
+                tags={"batch": "true"},
+            )
+            metrics.incr(
+                _INDEXER_DB_INSERT_ROWS_METRIC,
+                amount=len(rows_to_insert),
+                tags={"batch": "true"},
+            )
+            key_results.add_key_results(
+                [
+                    KeyResult(
+                        org_id=row.organization_id,
+                        string=row.string,
+                        id=row.decoded_id,
+                    )
+                    for row in rows_to_insert
+                ],
+                fetch_type=FetchType.FIRST_SEEN,
+            )
+
+    def _insert_individual_records(
+        self, rows_to_insert: Sequence[SpannerIndexerModel], key_results: KeyResults
+    ) -> None:
+        """
+        Insert the individual records in a transaction. We would need
+        to do this in scenarios where batch insert of the transaction
+        failed. Which can happen during collisions when multiple consumers
+        try to write records with same organization_id and string.
+
+        During an individual insert, we would get a unique constraint violation
+        exception for the existing record. The exception provides the
+        existing record id on which the record conflicts. We use this id
+        and decode it to populate the key_results.
+
+        decoded_id would be one of the following:
+        1. The id we generated for the record.
+        2. The id of the existing record on which the record conflicts
+        based on
+          a. The id received from the exception message when record was
+          being inserted.
+          b. The id of the existing record by performing a transaction read
+          for matching organization_id and string.
+
+        # TODO: Validate whether we should perform transaction read when
+        there is a conflict or we using the decoded it from the exception
+        is enough.
+        """
+
+        def insert_individual_record_uow(transaction: Any) -> None:
+            transaction.insert(self._table_name, columns=get_column_names(), values=[row])
+
+        metrics.incr(
+            _INDEXER_DB_INSERT_METRIC,
+            tags={"batch": "false"},
+        )
+        for row in rows_to_insert:
+            try:
+                self.database.run_in_transaction(insert_individual_record_uow)
```

Comment:
I think a better option would be to try a single read/write transaction here. The transaction reads all the rows you want to insert. See comment above. and try to insert what is supposed to not be there.

If we try the inserts one by one we are going to hit the whole transaction round trip for each record.
Moreover, relying on the read/write transaction removes the need of interpreting the error string in the response.

Reply:
Done